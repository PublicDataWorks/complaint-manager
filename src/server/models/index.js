const fs = require("fs");
const path = require("path");
const Sequelize = require("sequelize");
const basename = path.basename(__filename);
let env = process.env.NODE_ENV || "development";
if (env === "test") {
  env += process.env.JEST_WORKER_ID;
}
const config = require(__dirname + "/../config/sequelize_config.js")[env];
const db = {};
const dataChangeAuditHooks = require("../sequelizeHooks/dataChangeAuditHooks");
const caseStatusHooks = require("../sequelizeHooks/caseStatusHooks");

let sequelize;
if (config.use_env_variable) {
  sequelize = new Sequelize(process.env[config.use_env_variable], config);
} else {
  sequelize = new Sequelize(
    config.database,
    config.username,
    config.password,
    config
  );
}

//TODO Refactor to pass in only the class references we need.
dataChangeAuditHooks.init(sequelize, Sequelize.Model);
caseStatusHooks.init(sequelize, Sequelize.Model);

fs.readdirSync(__dirname)
  .filter(file => {
    return (
      file.indexOf(".") !== 0 && file !== basename && file.slice(-3) === ".js"
    );
  })
  .forEach(file => {
    var model = sequelize["import"](path.join(__dirname, file));
    db[model.name] = model;
  });

Object.keys(db).forEach(modelName => {
  if (db[modelName].associate) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
