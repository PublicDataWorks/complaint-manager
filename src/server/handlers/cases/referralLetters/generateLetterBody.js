import models from "../../../models";
import fs from "fs";
import Handlebars from "handlebars";

const getCaseData = async (caseId, transaction) => {
  return await models.cases.findById(caseId, {
    attributes: [
      ["id", "caseId"],
      "incidentDate",
      "incidentTime",
      "narrativeSummary",
      "narrativeDetails",
      "firstContactDate",
      "complaintType",
      "year",
      "caseNumber",
      "pibCaseNumber"
    ],
    order: [
      [
        { model: models.case_officer, as: "complainantOfficers" },
        "createdAt",
        "ASC"
      ],
      [
        { model: models.civilian, as: "complainantCivilians" },
        "createdAt",
        "ASC"
      ],
      [{ model: models.civilian, as: "witnessCivilians" }, "createdAt", "ASC"],
      [
        { model: models.case_officer, as: "witnessOfficers" },
        "createdAt",
        "ASC"
      ]
    ],
    include: [
      {
        model: models.referral_letter,
        as: "referralLetter",
        include: [
          {
            model: models.referral_letter_iapro_correction,
            as: "referralLetterIAProCorrections"
          }
        ]
      },
      { model: models.classification },
      {
        model: models.address,
        as: "incidentLocation"
      },
      {
        model: models.civilian,
        as: "complainantCivilians",
        include: [
          { model: models.address },
          { model: models.race_ethnicity, as: "raceEthnicity" }
        ]
      },
      {
        model: models.civilian,
        as: "witnessCivilians"
      },
      {
        model: models.case_officer,
        as: "complainantOfficers"
      },
      {
        model: models.case_officer,
        as: "accusedOfficers",
        separate: true,
        include: [
          {
            model: models.officer_allegation,
            as: "allegations",
            include: [{ model: models.allegation }]
          },
          {
            model: models.letter_officer,
            as: "letterOfficer",
            include: [
              {
                model: models.referral_letter_officer_history_note,
                as: "referralLetterOfficerHistoryNotes",
                separate: true
              },
              {
                model: models.referral_letter_officer_recommended_action,
                as: "referralLetterOfficerRecommendedActions",
                separate: true,
                include: [
                  {
                    model: models.recommended_action,
                    as: "recommendedAction"
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        model: models.case_officer,
        as: "witnessOfficers"
      }
    ],
    transaction
  });
};

async function generateLetterBody(caseId, transaction) {
  const caseData = (await getCaseData(caseId, transaction)).toJSON();
  caseData.accusedOfficers.sort((officerA, officerB) => {
    return officerA.createdAt > officerB.createdAt;
  });

  const rawTemplate = fs.readFileSync(
    "src/server/handlers/cases/referralLetters/getLetterPreview/letterBody.tpl"
  );

  const compiledTemplate = Handlebars.compile(rawTemplate.toString());

  return compiledTemplate(caseData);
}

export default generateLetterBody;
