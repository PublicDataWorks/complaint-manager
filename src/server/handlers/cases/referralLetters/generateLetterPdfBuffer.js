import models from "../../../policeDataManager/models";
import generatePdfBuffer from "./sharedLetterUtilities/generatePdfBuffer";
import Handlebars from "handlebars";
import getQueryAuditAccessDetails, {
  combineAuditDetails
} from "../../audits/getQueryAuditAccessDetails";
import { retrieveLetterImage } from "./retrieveLetterImage";
import { ASCENDING } from "../../../../sharedUtilities/constants";
require("../../../handlebarHelpers");

const MODEL_MAPPING = {
  complainantOfficers: { model: models.case_officer },
  complainantCivilians: { model: models.civilian },
  witnessCivilians: { model: models.civilian },
  witnessOfficers: { model: models.case_officer },
  referralLetter: { model: models.referral_letter },
  caseClassifications: { model: models.case_classification },
  classification: { model: models.classification },
  incidentLocation: { model: models.address },
  accusedOfficers: { model: models.case_officer, separate: true },
  allegations: { model: models.officer_allegation },
  allegation: { model: models.allegation },
  letterOfficer: { model: models.letter_officer },
  referralLetterOfficerHistoryNotes: {
    model: models.referral_letter_officer_history_note,
    separate: true
  },
  referralLetterOfficerRecommendedActions: {
    model: models.referral_letter_officer_recommended_action,
    separate: true
  },
  recommendedAction: { model: models.recommended_action },
  address: { model: models.address },
  raceEthnicity: { model: models.race_ethnicity },
  genderIdentity: { model: models.gender_identity }
};

const generateLetterPdfBuffer = async (
  caseId,
  includeSignature,
  transaction,
  letterSettings,
  extraData = {}
) => {
  let letterBody, auditDetails;

  const letterType = await models.letter_types.findOne({
    where: { type: letterSettings.type },
    include: ["fields"]
  });

  if (letterType.editableTemplate) {
    const queryOptions = {
      where: { caseId },
      attributes: ["editedLetterHtml"],
      transaction
    };
    let letterData = await models.referral_letter.findOne(queryOptions);
    letterBody = letterData.editedLetterHtml;

    ({ html: letterBody, auditDetails } = await determineLetterBody(
      letterBody,
      () =>
        getQueryAuditAccessDetails(queryOptions, models.referral_letter.name),
      letterType,
      caseId,
      transaction
    ));
  }

  const pdfDataAndAuditDetails = await getLetterData(
    caseId,
    letterType.fields.filter(field => !field.isForBody)
  );
  const pdfData = { ...pdfDataAndAuditDetails.data, ...extraData };
  const pdfDataAuditDetails = pdfDataAndAuditDetails.auditDetails;

  const fullLetterHtml = await generateLetterPdfHtml(
    letterBody,
    pdfData,
    includeSignature,
    letterSettings,
    letterType.template
  );

  auditDetails = auditDetails
    ? combineAuditDetails(auditDetails, pdfDataAuditDetails)
    : pdfDataAuditDetails;

  return {
    pdfBuffer: await generatePdfBuffer(fullLetterHtml),
    auditDetails: auditDetails
  };
};

export const generateLetterPdfHtml = async (
  letterBody,
  pdfData,
  includeSignature,
  letterSettings,
  template
) => {
  const currentDate = Date.now();

  let sender = pdfData.sender || pdfData.referralLetter?.sender;
  let signature = includeSignature
    ? await letterSettings.getSignature({ sender })
    : "<p><br></p>";
  let header = await retrieveLetterImage("header_text.png", "max-width: 223px");
  let smallIcon = await retrieveLetterImage("icon.ico", "max-width: 30px");
  let largeIcon = await retrieveLetterImage("icon.ico", "max-width: 42px");

  const letterPdfData = {
    ...pdfData,
    letterBody,
    signature,
    currentDate,
    header,
    smallIcon,
    largeIcon
  };

  const compiledTemplate = Handlebars.compile(template);
  return compiledTemplate(letterPdfData);
};

export const determineLetterBody = async (
  letterBody,
  auditIfEdited,
  letterType,
  caseId,
  transaction
) => {
  let auditDetails, html;
  if (letterBody) {
    html = letterBody;
    auditDetails = auditIfEdited();
  } else {
    const letterDataResults = await getLetterData(
      caseId,
      letterType.fields.filter(field => field.isForBody)
    );
    const compiledTemplate = Handlebars.compile(letterType.editableTemplate);
    html = compiledTemplate(letterDataResults.data);
    auditDetails = letterDataResults.auditDetails;
  }

  return { html, auditDetails };
};

export const getLetterData = async (caseId, fields) => {
  let queryOptions = {
    attributes: constructLetterDataQueryAttributes(fields),
    order: constructLetterDataQueryOrder(fields),
    include: constructLetterDataQueryInclude(fields)
  };

  let letterData = await models.cases.findByPk(caseId, queryOptions);

  letterData = letterData.toJSON();
  if (letterData?.accusedOfficers) {
    letterData.accusedOfficers = letterData.accusedOfficers.sort(
      (officerA, officerB) => {
        return officerA.createdAt - officerB.createdAt;
      }
    );
  }

  return {
    data: letterData,
    auditDetails: getQueryAuditAccessDetails(queryOptions, models.cases.name)
  };
};

const constructLetterDataQueryInclude = fields => {
  return fields
    .filter(field => field.relation !== "cases")
    .reduce(
      (include, f) =>
        addRelationToIncludeClause(f.relation.split("."), include, f.field),
      []
    );
};

const constructLetterDataQueryOrder = fields => {
  return fields
    .filter(field => !!field.sortBy)
    .map(field => {
      let relationName = field.relation.split(".").pop();
      return [
        { model: MODEL_MAPPING[relationName].model, as: relationName },
        field.sortBy,
        field.sortDirection || ASCENDING
      ];
    });
};

const constructLetterDataQueryAttributes = fields => {
  return fields.filter(field => field.relation === "cases").map(f => f.field);
};

/**
 * addRelationToIncludeClause - recursive function that creates sequelize "include" objects
 * and adds them to a given array (to be used in a sequelize query)
 *
 * @param {Array} pathToRelation an array of strings that shows the nesting needed
 * to place the field at the right depth
 * @param {Array} placeToAddRelation the include array that has been constructed so far (expect to pass a reference to an empty array on the initial call)
 * @param {String} field the name of the attribute to be included (if * then all fields should be included)
 * @returns the array (originally placeToAddRelation) that contains the constructed include clause(s)
 */
const addRelationToIncludeClause = (
  pathToRelation,
  placeToAddRelation,
  field
) => {
  let relationName = pathToRelation.shift();

  let existingObject = placeToAddRelation.find(obj => obj.as === relationName);
  if (!existingObject) {
    existingObject = createIncludeObject(relationName, placeToAddRelation);
  }

  if (pathToRelation.length) {
    existingObject.include = existingObject.include || [];
    addRelationToIncludeClause(pathToRelation, existingObject.include, field);
  } else {
    // base case
    if (field !== "*") {
      existingObject.attributes = existingObject.attributes
        ? existingObject.attributes
        : [];
      existingObject.attributes.push(field);
    }
  }

  return placeToAddRelation;
};

const createIncludeObject = (relationName, placeToAddRelation) => {
  let existingObject = {
    model: MODEL_MAPPING[relationName].model
  };

  if (relationName !== "address" && relationName !== "allegation") {
    existingObject.as = relationName;
  }

  if (MODEL_MAPPING[relationName].separate) {
    existingObject.separate = true;
  }

  placeToAddRelation.push(existingObject);
  return existingObject;
};

export default generateLetterPdfBuffer;
