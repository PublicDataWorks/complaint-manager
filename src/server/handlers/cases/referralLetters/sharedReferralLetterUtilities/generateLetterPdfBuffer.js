import models from "../../../../models/index";
import generateLetterBody from "../generateLetterBody";
import pdf from "html-pdf";
import util from "util";
import Handlebars from "handlebars";
require("../../../../handlebarHelpers");
import fs from "fs";

const generateLetterPdfBuffer = async (
  caseId,
  includeSignature,
  transaction
) => {
  let letterData = await models.referral_letter.find({
    where: { caseId: caseId },
    attributes: ["editedLetterHtml"],
    transaction
  });
  let letterBody = letterData.editedLetterHtml;

  if (!letterBody) {
    letterBody = await generateLetterBody(caseId, transaction);
  }

  const pdfData = await getPdfData(caseId, transaction);

  const pdfOptions = {
    format: "Letter",
    timeout: 100000,
    width: "8.5in",
    height: "11in",
    border: "0.5in",
    header: { height: "1.3 in" },
    footer: { height: "0.7 in" },
    base: "file:///app/src/server/handlers/cases/referralLetters/getPdf/assets/"
  };

  const fullLetterHtml = await generateLetterPdfHtml(
    letterBody,
    pdfData,
    includeSignature
  );

  let pdfCreator = pdf.create(fullLetterHtml, pdfOptions);
  let pdfToBuffer = util.promisify(pdfCreator.toBuffer.bind(pdfCreator));

  return await pdfToBuffer();
};

const getPdfData = async (caseId, transaction) => {
  return await models.cases.findById(caseId, {
    attributes: [
      "firstContactDate",
      "complaintType",
      "id",
      "year",
      "caseNumber",
      "pibCaseNumber"
    ],
    include: [
      {
        model: models.referral_letter,
        as: "referralLetter",
        attributes: ["recipient", "sender", "transcribedBy"]
      }
    ],
    transaction
  });
};

export const generateLetterPdfHtml = (
  letterBody,
  pdfData,
  includeSignature
) => {
  const currentDate = Date.now();

  const letterPdfData = {
    letterBody: letterBody,
    recipient: pdfData.referralLetter.recipient,
    sender: pdfData.referralLetter.sender,
    transcribedBy: pdfData.referralLetter.transcribedBy,
    caseReference: pdfData.caseReference,
    pibCaseNumber: pdfData.pibCaseNumber,
    includeSignature,
    currentDate
  };

  const rawTemplate = fs.readFileSync(
    "src/server/handlers/cases/referralLetters/getPdf/letterPdf.tpl"
  );
  const compiledTemplate = Handlebars.compile(rawTemplate.toString());
  return compiledTemplate(letterPdfData);
};

export default generateLetterPdfBuffer;
