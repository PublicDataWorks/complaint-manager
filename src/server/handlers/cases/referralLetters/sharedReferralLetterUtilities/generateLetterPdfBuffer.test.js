import { generateLetterPdfHtml } from "./generateLetterPdfBuffer";
import timekeeper from "timekeeper";
import Case from "../../../../../client/testUtilities/case";
import ReferralLetter from "../../../../../client/testUtilities/ReferralLetter";
import { cleanupDatabase } from "../../../../testHelpers/requestTestHelpers";
import generateLetterBody from "../generateLetterBody";
import models from "../../../../models";
import {
  CASE_STATUS,
  CIVILIAN_INITIATED
} from "../../../../../sharedUtilities/constants";
import generateLetterPdfBuffer from "./generateLetterPdfBuffer";

jest.mock("html-pdf", () => ({
  create: (html, pdfOptions) => ({
    toBuffer: callback => {
      callback(null, html);
    }
  })
}));

jest.mock("../generateLetterBody");

describe("generateLetterPdfBuffer", () => {
  let existingCase, referralLetter, timeOfDownload;

  afterEach(async () => {
    await cleanupDatabase();
    generateLetterBody.mockReset();
    timekeeper.reset();
  });

  beforeEach(async () => {
    timeOfDownload = new Date("2018-07-01 19:00:22 CDT");
    timekeeper.freeze(timeOfDownload);
    const caseAttributes = new Case.Builder()
      .defaultCase()
      .withId(12070)
      .withFirstContactDate("2017-12-25")
      .withIncidentDate("2016-01-01")
      .withComplaintType(CIVILIAN_INITIATED);
    existingCase = await models.cases.create(caseAttributes, {
      auditUser: "test"
    });
    await existingCase.update(
      { status: CASE_STATUS.ACTIVE },
      { auditUser: "test" }
    );
    await existingCase.update(
      { status: CASE_STATUS.LETTER_IN_PROGRESS },
      { auditUser: "test" }
    );
  });

  describe("sender is 'sender address'", async () => {
    beforeEach(async () => {
      const referralLetterAttributes = new ReferralLetter.Builder()
        .defaultReferralLetter()
        .withId(undefined)
        .withCaseId(existingCase.id)
        .withRecipient("recipient address")
        .withSender("sender address")
        .withTranscribedBy("transcriber")
        .withIncludeRetaliationConcerns(true);

      referralLetter = await models.referral_letter.create(
        referralLetterAttributes,
        {
          auditUser: "test"
        }
      );
    });
    test("generates letter pdf html correctly", async () => {
      const letterBody = "<p> Letter Body </p>";
      const pdfData = {
        referralLetter: {
          recipient: "Recipient Address",
          sender: "Sender Address\n Sender Address Second Line",
          transcribedBy: "Transcriber"
        },
        caseReference: "CC-2011-0099"
      };

      const letterPdfHtml = await generateLetterPdfHtml(
        letterBody,
        pdfData,
        false
      );
      expect(letterPdfHtml).toMatchSnapshot();
    });

    test("pdf create gets called with expected letter html when letter is generated", async () => {
      await models.sequelize.transaction(async transaction => {
        const pdfResults = await generateLetterPdfBuffer(
          existingCase.id,
          true,
          transaction
        );
        expect(pdfResults).toMatchSnapshot();
      });
    });

    test("pdf create gets called with expected letter html when letter is edited", async () => {
      await referralLetter.update(
        { editedLetterHtml: "Custom Letter HTML" },
        { auditUser: "someone" }
      );

      await models.sequelize.transaction(async transaction => {
        const pdfResults = await generateLetterPdfBuffer(
          existingCase.id,
          true,
          transaction
        );
        expect(pdfResults).toMatchSnapshot();
      });
    });

    test("signature is not included when sender is not stella even if includeSignature is true", async () => {
      await referralLetter.update(
        { editedLetterHtml: "Custom Letter HTML" },
        { auditUser: "someone" }
      );
      await models.sequelize.transaction(async transaction => {
        const pdfResults = await generateLetterPdfBuffer(
          existingCase.id,
          true,
          transaction
        );
        expect(pdfResults).toMatchSnapshot();
      });
    });

    test("unedited letter generates pdf from case data", async () => {
      await models.sequelize.transaction(async transaction => {
        await generateLetterPdfBuffer(existingCase.id, false, transaction);
        expect(generateLetterBody).toHaveBeenCalledWith(
          existingCase.id,
          transaction
        );
      });
    });

    test("edited letter generates pdf from saved data", async () => {
      await referralLetter.update(
        {
          editedLetterHtml: "<p> edited </p>"
        },
        { auditUser: "test" }
      );
      await models.sequelize.transaction(async transaction => {
        await generateLetterPdfBuffer(existingCase.id, false, transaction);
      });
      expect(generateLetterBody).not.toHaveBeenCalled();
    });
  });

  describe("sender is stella", async () => {
    beforeEach(async () => {
      const referralLetterAttributes = new ReferralLetter.Builder()
        .defaultReferralLetter()
        .withId(undefined)
        .withCaseId(existingCase.id)
        .withRecipient("recipient address")
        .withSender("Stella Cziment")
        .withTranscribedBy("transcriber")
        .withIncludeRetaliationConcerns(true);

      referralLetter = await models.referral_letter.create(
        referralLetterAttributes,
        {
          auditUser: "test"
        }
      );
    });

    test("signature is included when sender is stella and includeSignature is true", async () => {
      await models.sequelize.transaction(async transaction => {
        const pdfResults = await generateLetterPdfBuffer(
          existingCase.id,
          true,
          transaction
        );
        expect(pdfResults).toMatchSnapshot();
      });
    });

    test("signature is not included when includeSignature is false", async () => {
      await models.sequelize.transaction(async transaction => {
        const pdfResults = await generateLetterPdfBuffer(
          existingCase.id,
          false,
          transaction
        );
        expect(pdfResults).toMatchSnapshot();
      });
    });
  });
});
