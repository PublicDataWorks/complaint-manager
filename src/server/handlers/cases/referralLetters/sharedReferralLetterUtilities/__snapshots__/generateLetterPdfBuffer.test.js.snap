// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateLetterPdfBuffer sender is 'sender address' generates letter pdf html correctly 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      Recipient Address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    Recipient Address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC-2011-0099</strong></p>
    <p><br></p>
    <p>Dear Recipient Address:</p>
    <p><br></p>
    <p> Letter Body </p>
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    Sender Address<br> Sender Address Second Line
    <p><br></p>
    Transcribed by: Transcriber
  </body>
</html>"
`;

exports[`generateLetterPdfBuffer sender is 'sender address' pdf create gets called with expected letter html when letter is edited 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      recipient address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    recipient address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC2017-0001</strong></p>
    <p><br></p>
    <p>Dear recipient address:</p>
    <p><br></p>
    Custom Letter HTML
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    sender address
    <p><br></p>
    Transcribed by: transcriber
  </body>
</html>"
`;

exports[`generateLetterPdfBuffer sender is 'sender address' pdf create gets called with expected letter html when letter is generated 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      recipient address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    recipient address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC2017-0001</strong></p>
    <p><br></p>
    <p>Dear recipient address:</p>
    <p><br></p>
    
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    sender address
    <p><br></p>
    Transcribed by: transcriber
  </body>
</html>"
`;

exports[`generateLetterPdfBuffer sender is 'sender address' signature is not included when sender is not stella even if includeSignature is true 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      recipient address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    recipient address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC2017-0001</strong></p>
    <p><br></p>
    <p>Dear recipient address:</p>
    <p><br></p>
    Custom Letter HTML
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    sender address
    <p><br></p>
    Transcribed by: transcriber
  </body>
</html>"
`;

exports[`generateLetterPdfBuffer sender is stella signature is included when sender is stella and includeSignature is true 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      recipient address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    recipient address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC2017-0001</strong></p>
    <p><br></p>
    <p>Dear recipient address:</p>
    <p><br></p>
    
    <p><br></p>
    Sincerely,
    <p><br></p>
    <img style=\\"max-height: 55px\\" src=file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/signatures/signature.png />
    <p><br></p>
    Stella Cziment
    <p><br></p>
    Transcribed by: transcriber
  </body>
</html>"
`;

exports[`generateLetterPdfBuffer sender is stella signature is not included when includeSignature is false 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"getPdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/header_text.png\\" />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      recipient address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=\\"file:/app/src/server/handlers/cases/referralLetters/getPdf/assets/icon.ico\\" />
            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    recipient address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint CC2017-0001</strong></p>
    <p><br></p>
    <p>Dear recipient address:</p>
    <p><br></p>
    
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    Stella Cziment
    <p><br></p>
    Transcribed by: transcriber
  </body>
</html>"
`;
